ptp/components/models/vqa/multimodal_factorized_bilinear_pooling.py

```diff
@@ -0,0 +1,126 @@
+
+        # Get size of latent space and number of heads from config.
+        self.latent_size = self.config["latent_size"]
+        self.factor = self.config["pool_factor"]
+        # Output feature size
+        self.output_size = self.latent_size
+
+        # Map image and question encodings to a common latent space of dimension 'latent_size'.
+        self.image_encodings_ff = torch.nn.Linear(self.image_encoding_size, self.latent_size*self.factor)
+        self.question_encodings_ff = torch.nn.Linear(self.question_encoding_size, self.latent_size*self.factor)
+
+        # Create activation layer.
+        self.activation = torch.nn.ReLU()
+
+        # Retrieve dropout rate value - if set, will put dropout between every layer.
+        dropout_rate = self.config["dropout_rate"]
+
+        # Create dropout layer.
+        self.dropout = torch.nn.Dropout(dropout_rate)
+
+
+    def input_data_definitions(self):
+        """
+        Function returns a dictionary with definitions of input data that are required by the component.
+
+        :return: dictionary containing input data definitions (each of type :py:class:`ptp.utils.DataDefinition`).
+        """
+        return {
+            self.key_image_encodings: DataDefinition([-1, self.image_encoding_size], [torch.Tensor], "Batch of encoded images [BATCH_SIZE x IMAGE_ENCODING_SIZE]"),
+            self.key_question_encodings: DataDefinition([-1, self.question_encoding_size], [torch.Tensor], "Batch of encoded questions [BATCH_SIZE x QUESTION_ENCODING_SIZE]"),
+            }
+
+
+    def output_data_definitions(self):
+        """
+        Function returns a dictionary with definitions of output data produced the component.
+
+        :return: dictionary containing output data definitions (each of type :py:class:`ptp.utils.DataDefinition`).
+        """
+        return {
+            self.key_outputs: DataDefinition([-1, self.output_size], [torch.Tensor], "Batch of outputs [BATCH_SIZE x OUTPUT_SIZE]")
+            }
+
+    def forward(self, data_dict):
+        """
+        Main forward pass of the model.
+
+        :param data_dict: DataDict({'images',**})
+        :type data_dict: ``ptp.dadatypes.DataDict``
+        """
+
+        # Unpack DataDict.
+        enc_img = data_dict[self.key_image_encodings] #[48, 2048]
+        enc_q = data_dict[self.key_question_encodings] #[48, 100]
+
+        # Map image and question encodings to high-dimensional space using FF
+        latent_img = self.dropout(self.image_encodings_ff(enc_img)) # [48, 512]
+        latent_q =  self.dropout(self.question_encodings_ff(enc_q)) # [48, 512]
+
+        # Element-wise mutliplication of image and question encodings
+        enc_z = latent_img * latent_q # [48, 512]
+        # Dropout regularization
+        enc_z = self.dropout(enc_z)
+        enc_z = enc_z.view(enc_z.size(0), self.latent_size, self.factor) # [48, 256, 2]
+        # Sum pooling
+        enc_z = enc_z.sum(2) # [48, 256]
```

Comment:
in the paper they eq. (4) introduces k=2, if I am right this "k" is self.factor ("pooling_factor").

Should that be used in here?

Reply:
would it be better to change line 120 to
enc_z = torch.sum(enc_z, dim=self.factor) ?

Reply:
you are right! :)